ProcessTree.cpp

```diff
@@ -0,0 +1,525 @@
+            if (_processes.count(rit->pid) > 0) {
+                process->_flags = _filtersEngine->GetFlags(_processes[rit->pid], height);
+            }
+        }
+    }
+}
+
+
+void ProcessTree::PopulateTree()
+{ 
+    struct dirent *de;
+    int pid;
+
+    DIR *dr = opendir("/proc"); 
+  
+    if (dr == NULL)
+    { 
+        return;
+    } 
+  
+    while ((de = readdir(dr)) != NULL)  {
+        if (is_number(de->d_name)) {
+            pid = atoi(de->d_name);
+            auto process = ReadProcEntry(pid);
+            if (process != nullptr) {
+                _processes[pid] = process;
+            }
+        }
+    }
+
+    closedir(dr);     
+
+    for (auto p : _processes) {
+        auto process = p.second;
+        if (_processes.count(process->_ppid) > 0) {
+            _processes[process->_ppid]->_children.emplace_back(process->_pid);
+        }
+    }
+
+    for (auto p : _processes) {
+        std::shared_ptr<ProcessTreeItem> process, parent;
+        process = p.second;
+        if (_processes.count(process->_ppid) > 0) {
+            parent = _processes[process->_ppid];
+        } else {
+            parent = nullptr;
+        }
+        while (parent) {
+            process->_ancestors.insert(process->_ancestors.begin(), {parent->_pid, parent->_exe});
+            if (_processes.count(parent->_ppid) > 0) {
+                parent = _processes[parent->_ppid];
+            } else {
+                parent = nullptr;
+            }
+        }
+    }
+
+    for (auto p : _processes) {
+        ApplyFlags(p.second);
+    }
+} 
+
+std::string ProcessTree::ReadFirstLine(const std::string& file)
+{
+    std::ifstream f;
+    std::string line;
+
+    try {
+        f.open(file);
+        std::getline(f, line);
+        f.close();
+    } catch (...) {
+    }
+    return line;
+}
+
+std::string ProcessTree::ReadParam(const std::string& file, const std::string& param)
+{
+    std::ifstream f;
+    std::string line;
+    std::string value;
+
+    f.open(file);
+    while (f.good() && f.is_open() && !f.eof()) {
+        std::getline(f, line);
+        if (!line.compare(0, param.size() + 1, param + ":")) {
+            f.close();
+            value = line.substr(param.size() + 1);
+            size_t first = value.find_first_not_of(" \t");
+            if ( first == std::string::npos) {
+                return value;
+            } else {
+                return value.substr(first);
+            }
+        }
+    }
+    f.close();
+    return value;
+}
+
+std::shared_ptr<ProcessTreeItem> ProcessTree::ReadProcEntry(int pid)
+{
+    std::shared_ptr<ProcessTreeItem> process = std::make_shared<ProcessTreeItem>(ProcessTreeSource_procfs, pid);
+    std::string prefix = std::string("/proc/") + std::to_string(pid) + "/";
+
+    // Check if the pid dir exists
+    DIR *dir = opendir(prefix.c_str());
+    if (!dir) {
+        return nullptr;
+    }
+
+    try {
+        std::string uidline = ReadParam(prefix + "status", "Uid");
```

Comment:
ReadParam() is called three times here. Each time the status file is opened. This is at least an order of magnitude slower the same log in ProcessInfo.cpp. Please consider using ProcessInfo instead. You could modify ProcessInfo and add an option to only read/parse the status file.

Reply:
As status is a virtual file this won't have quite the same impact as if this was a physical file. Also, other than on start up, this should be a particularly rare event so I'm less concerned about performance here. It could certainly be looked at when the rest of the code base is reviewed.

Reply:
Replaced with ProcessInfo::Open() as suggested.